trpcultivate_phenotypes/tests/src/Unit/ConfigRRulesFormTest.php

```diff
@@ -0,0 +1,145 @@
+
+use Drupal\trpcultivate_phenotypes\Form\TripalCultivatePhenotypesRSettingsForm;
+use Drupal\Core\Config\Config;
+use Drupal\Core\Form\FormState;
+use Drupal\Tests\UnitTestCase;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\StringTranslation\TranslationInterface;
+use Drupal\Core\DependencyInjection\ContainerBuilder;
+
+ /**
+  *  Class definition ConfigRRulesFormTest.
+  *
+  * @coversDefaultClass Drupal\trpcultivate_phenotypes\Form\TripalCultivatePhenotypesRSettingsForm
+  * @group trpcultivate_phenotypes
+  */
+class ConfigRRulesFormTest extends UnitTestCase {
+  
+  /**
+   * Initialization of container, configurations, service 
+   * and service class required by the test.
+   */
+  protected function setUp(): void {
+    parent::setUp();
+    
+    // Create container.
+    $container = new ContainerBuilder();
+    \Drupal::setContainer($container);
+
+    // Mock config since the RRules form extends configFormBase and expects
+    // a configuration settings.
+    $r_config_mock = $this->prophesize(Config::class);
+    $r_config_mock->get('trpcultivate.phenotypes.r_config')->willReturn([
+      'words' => ['of', 'to', 'have', 'on', 'at'],
+      'chars' => ['(', ')', '/', '-', ':', ';', '%'],
+      'replace' => [ '# = num', '/ = div', '? = unsure', '- = to']
+    ]);
+   
+    // When RRules form rebuilds calling the module settings, return
+    // only the R configuration settings above exclude other config.
+    $all_config_mock = $this->prophesize(ConfigFactoryInterface::class);
+    $all_config_mock->getEditable('trpcultivate_phenotypes.settings')
+      ->willReturn($r_config_mock);
+
+    // Isolated configuration for R Rules.
+    $r_config = $all_config_mock->reveal();
+
+    // Translation requirement of the container
+    $translation_mock = $this->prophesize(TranslationInterface::class);
+    $translation = $translation_mock->reveal();
+
+    // Class RRulesForm class instance.
+    $rrules_form = new TripalCultivatePhenotypesRSettingsForm($r_config);
+    $rrules_form->setStringTranslation($translation);
+
+    $container->set('rrules.config', $rrules_form);
+  }
+
+  /**
+   * Test validate functionality of RRulesForm class.
+   */  
+  public function testValidateForm() {
+    $rrules = \Drupal::service('rrules.config');
+
+    // Test if it is the RRules config form using the form id.
+    $this->assertEquals('trpcultivate_phenotypes_r_settings_form', $rrules->getFormId());
+    
+    // Method formValidate requires 2 parameters $form and $form_state.
+    $form_state = new FormState();
+    $form = [];
+    
+    // Validation: WORDS Rule
+    // words - any words at least 2 characters long and no empty value.
+    // Failed, Has validation error:
+    $field = 'words';
+    foreach(['R', 'r', '.', '~', '1', '123', ' '] as $rule) {
+      $form_state->setValue($field, $rule);
+      $rrules->validateForm($form, $form_state);
+
+      $this->assertTrue($form_state->hasAnyErrors(), $rule);
```

Comment:
I think you need the `$form_state->clearErrors();` inside this loop 🤔 

I know you're setting the same field but I believe the errors may just be building up. You can check this by printing the errors in the form state to the screen after the loop completes.

Reply:
Part of the recent commit included passing unit test for buildForm and formValidate. The issue with submit was it could save configuration because it could not retrieve the configuration name. I will revisit this issue.